service-base/src/main/java/org/eclipse/hono/service/registration/AbstractRegistrationService.java

```diff
@@ -216,22 +237,49 @@ protected boolean isGatewayAuthorized(final String gatewayId, final JsonObject g
         Objects.requireNonNull(deviceId);
         Objects.requireNonNull(deviceData);
 
-        final Object obj = deviceData.getValue(RegistrationConstants.FIELD_VIA);
-        if (obj instanceof JsonArray) {
-            // get the first matching entry
-            return ((JsonArray) obj)
-                    .stream()
-                    .filter(String.class::isInstance)
-                    .anyMatch(gatewayId::equals);
-        } else if (obj instanceof String) {
-            // compare the string directly
-            return gatewayId.equals(obj);
+        final Object via = deviceData.getValue(RegistrationConstants.FIELD_VIA);
+        final Object memberOf = gatewayData.getValue(RegistrationConstants.FIELD_MEMBER_OF);
+
+        if (via instanceof JsonArray && memberOf instanceof JsonArray) {
+            return isGatewayInVia(gatewayId, (JsonArray) via) ||
+                    isGatewayGroupInVia((JsonArray) memberOf, (JsonArray) via);
+        } else if (via instanceof String && memberOf instanceof JsonArray) {
+            return gatewayId.equals(via) || ((JsonArray) memberOf).contains(via);
+        } else if (via instanceof JsonArray && memberOf instanceof String) {
+            return isGatewayInVia(gatewayId, (JsonArray) via) || isGatewayGroupInVia((String) memberOf, (JsonArray) via);
+        } else if (via instanceof String && memberOf instanceof String) {
+            return gatewayId.equals(via) || memberOf.equals(via);
+        } else if (via instanceof JsonArray && memberOf == null) {
+            return isGatewayInVia(gatewayId, (JsonArray) via);
+        } else if (via instanceof String && memberOf == null) {
+            return gatewayId.equals(via);
```

Comment:
I think working only with "via" and "memberOf" (Json-) arrays (creating a single element array in case of a String value or an empty array in case of an empty "memberOf") would simplify the above block quite a bit.